core/src/layers/madsim.rs

```diff
@@ -0,0 +1,420 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::Result;
+use std::io::SeekFrom;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::{Context, Poll};
+
+/// Add deterministic simulation for async operations, powered by [`madsim`](https://docs.rs/madsim/latest/madsim/).
+///
+/// # Note
+///
+/// - blocking operations are not supported, as [`madsim`](https://docs.rs/madsim/latest/madsim/) is async only.
+///
+///
+/// # Examples
+///
+/// ```
+/// use opendal::Operator;
+/// use opendal::services;
+/// use opendal::layers::MadsimLayer;
+/// use opendal::layers::MadsimServer;
+/// use madsim::{net::NetSim, runtime::Handle, time::sleep};
+/// use std::time::Duration;
+///
+/// #[cfg(madsim)]
+/// #[madsim::test]
+/// async fn deterministic_simulation_test(){
+///     let handle = Handle::current();
+///     let ip1 = "10.0.0.1".parse().unwrap();
+///     let ip2 = "10.0.0.2".parse().unwrap();
+///     let sim_server_socket = "10.0.0.1:2379".parse().unwrap();
+///     let server = handle.create_node().name("server").ip(ip1).build();
+///     let client = handle.create_node().name("client").ip(ip2).build();
+///
+///     server.spawn(async move {
+///          SimServer::serve(sim_server_socket).await.unwrap();
+///     });
+///     sleep(Duration::from_secs(1)).await;
+///
+///     let handle = client.spawn(async move {
+///     let mut builder = services::Fs::default();
+///     builder.root(".");
+///     let op = Operator::new(builder)
+///         .unwrap()
+///         .layer(MadsimLayer::new(sim_server_socket))
+///         .finish();
+///
+///          let path = "hello.txt";
+///          let data = "Hello, World!";
+///          op.write(path, data).await.unwrap();
+///          assert_eq!(data.as_bytes(), op.read(path).await.unwrap());
+///      });
+///      handle.await.unwrap();
+/// }
+/// ```
+/// To enable logging output, please set `RUSTFLAGS="--cfg madsim"`:
+/// ```shell
+/// RUSTFLAGS="--cfg madsim" cargo test
+/// ```
+#[derive(Debug, Copy, Clone)]
+pub struct MadsimLayer {
+    addr: SocketAddr,
+}
+
+impl MadsimLayer {
+    pub fn new(endpoint: &str) -> Self {
+        Self {
+            addr: endpoint.parse().unwrap(),
+        }
+    }
+}
+
+impl<A: Accessor> Layer<A> for MadsimLayer {
+    type LayeredAccessor = MadsimAccessor;
+
+    fn layer(&self, _: A) -> Self::LayeredAccessor {
+        MadsimAccessor { addr: self.addr }
+    }
+}
+
+#[derive(Debug)]
+pub struct MadsimAccessor {
+    addr: SocketAddr,
+}
+
+#[async_trait]
+impl LayeredAccessor for MadsimAccessor {
+    type Inner = ();
+    type Reader = MadsimReader;
+    type BlockingReader = ();
+    type Writer = MadsimWriter;
+    type BlockingWriter = ();
+    type Pager = MadsimPager;
+    type BlockingPager = ();
+
+    fn inner(&self) -> &Self::Inner {
+        &()
+    }
+
+    fn metadata(&self) -> AccessorInfo {
+        let mut info = AccessorInfo::default();
+        info.set_name("madsim");
```

Comment:
Please also set [`AccessorCapability`](https://docs.rs/opendal/latest/opendal/raw/enum.AccessorCapability.html)

Based on current implementation, we can set `Read | Write`.